raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -814,6 +822,10 @@ private EndQuorumEpochResponseData handleEndQuorumEpochRequest(
     ) throws IOException {
         EndQuorumEpochRequestData request = (EndQuorumEpochRequestData) requestMetadata.data;
 
+        if (!hasValidClusterId(request)) {
```

Comment:
I think a better way to do this is to modify `validateVoterOnlyRequest` and `validateLeaderOnlyRequest` so that we pass the clusterId. Then we can get rid of `getClusterId`.

Reply:
I tried this approach, it seems that voter and leader is partition level terminology so that `validateVoterOnlyRequest` is used to get a partition level error but cluster validate is a request level error, we'd better separate these 2 errors since we are making way for multi-raft. I think we can add a `validateRequest` method to get request level error, WDYT.